Update to govuk-frontend 3.0

alex-ju opened the pull request:
## What
Update `govuk_publishing_components` to use `govuk-frontend` 3.0.
Fix focus states on a few components to align them with the colour contrast requirements

## Why
- Keep frontend code in sync with the Design System updates
- Meet [WCAG 2.1 requirements regarding colour contrast](https://www.w3.org/TR/WCAG21/#contrast-minimum)

## Notes
- this PR is ~~raised for visibility and be used as a base for discussions~~ ready for review
- for more details regarding changes see [`govuk-frontend` 3.0 changelog](https://github.com/alphagov/govuk-frontend/releases/tag/v3.0.0)

## Compatibility checks
These changes were tested with applications in the following states using the appropriate compatibility flags.

- [x] applications with a direct dependency on `govuk_elements_rails`, `govuk_frontend_toolkit` and an indirect dependency on `govuk_template` (e.g. Service manual frontend)
- [x] applications with a direct dependency on `govuk_frontend_toolkit` and an indirect dependency on `govuk_template` (e.g. Frontend)
- [x] applications with an indirect dependency on `govuk_template` - dependent on static (e.g. [Calendars](https://github.com/alphagov/calendars/compare/update-to-govuk-frontend-3-0))
- [x] applications without dependencies on legacy projects (e.g. [Content Publisher](https://github.com/alphagov/content-publisher/compare/update-to-govuk-frontend-3-0))


## Visual Changes
There are visual changes for almost all components, especially related to [focus states](https://design-system.service.gov.uk/get-started/focus-states/) (which are harder to get with a basic visual comparison). I would recommend using the preview app when reviewing as images don't capture all the states.

## View Changes
https://govuk-publishing-compo-pr-1010.herokuapp.com/component-guide

[Trello card](https://trello.com/c/bYtf5EEe)

injms commented on app/assets/javascripts/govuk_publishing_components/lib/header-navigation.js:
```diff
@@ -4,7 +4,7 @@
+    var els = document.querySelectorAll('.js-header-toggle')
     var i
     var _i
     for (i = 0, _i = els.length; i < _i; i++) {
```
I'm curious and intrigued about this loop, since it seems to assigned `els.length` to `_i` within parenthesis, and I wonder what this does and why?

alex-ju replied:
🤷‍♂️haven't looked at it to be honest. Hope this will soon be removed as we have it covered by the header script from govuk-frontend.

36degrees commented on app/assets/stylesheets/govuk_publishing_components/components/_attachment.scss:
```diff
@@ -5,7 +5,7 @@ $thumbnail-background: govuk-colour("white");
 $thumbnail-shadow-colour: rgba(11, 12, 12, .4);
 $thumbnail-shadow-width: 0 2px 2px;
-$thumbnail-icon-border-colour: govuk-colour("grey-3");
+$thumbnail-icon-border-colour: govuk-colour("mid-grey");
```
If you want these components to work in downstream applications with compatibility mode enabled, you'll want to pass `$legacy: "grey-3"` here so that it knows which colour to use from the legacy palette, whenever you use a colour that's been renamed or added.

Without this, if you tried to use the attachment component in an app that has compatibility mode enabled, you'd get an 'Unknown colour mid-grey' error.

http://govuk-frontend-review.herokuapp.com/docs/#function-govuk-colour

You might also want to find a way to surface or test with legacy mode within publishing components, so that you don't find out things like this are broken after you've published!

alex-ju replied:
Sounds good. Wasn't aware of the `$legacy` attribute - haven't seen it anywhere yet.

alex-ju replied:
Adressed in 4c479d7

36degrees commented on app/assets/stylesheets/component_guide/application.scss:
```diff
@@ -2,12 +2,12 @@
+@import "govuk/core/all";
 
-@include _govuk-font-face-nta;
+@include _govuk-font-face-gds-transport;
```
As an aside, the font-face should be included automatically the first time it's used – you shouldn't have to do this. If that's not working for you, it'd be good to know, as we consider this a private mixin at the minute.

alex-ju replied:
This should've been removed after dropping the compatibility mode.

alex-ju replied:
Removed

36degrees commented on app/assets/stylesheets/govuk_publishing_components/components/_highlight-boxes.scss:
```diff
@@ -66,7 +66,7 @@
   &:active,
   &:focus {
-    color: $govuk-link-colour;
+    color: $govuk-text-colour;
```
Unsure, because of the `:active` state, but should this be `$govuk-focus-text-colour`?

alex-ju replied:
Yep, they all (for all the fixed focus colours) should be `$govuk-focus-text-colour`

alex-ju replied:
Checked focus states for all components and made use of `$govuk-focus-text-colour` where needed in ff411ec

36degrees commented on app/assets/stylesheets/govuk_publishing_components/components/_input.scss:
```diff
@@ -1,4 +1,4 @@
+@import "govuk/components/input/input";
 
 .gem-c-input--search-icon {
   background: govuk-colour("white") url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 36 36' width='40' height='40'%3E%3Cpath d='M25.7 24.8L21.9 21c.7-1 1.1-2.2 1.1-3.5 0-3.6-2.9-6.5-6.5-6.5S10 13.9 10 17.5s2.9 6.5 6.5 6.5c1.6 0 3-.6 4.1-1.5l3.7 3.7 1.4-1.4zM12 17.5c0-2.5 2-4.5 4.5-4.5s4.5 2 4.5 4.5-2 4.5-4.5 4.5-4.5-2-4.5-4.5z' fill='currentColor'%3E%3C/path%3E%3C/sv ...
```
Unrelated to this change, and super minor, but as far as I know `currentColor` doesn't work in background images – so the magnifying glass icon is `#000000`

alex-ju replied:
Good call for improvement!

36degrees commented on app/assets/stylesheets/govuk_publishing_components/components/_feedback.scss:
```diff
@@ -33,10 +33,8 @@
-
 .gem-c-feedback__column-two-thirds {
-  @include govuk-grid-column(two-thirds, $class: false)
+  @include govuk-grid-column($width: two-thirds)
```
Could the feedback component just be updated to use `govuk-grid-column-two-thirds`? It looks like the row class was changed to `govuk-grid-row` in the markup.

alex-ju replied:
Yes, wasn't constrained to do it for the row, but still got wiggle room for columns. I personally think it should use both classes from `govuk-frontend`.

alex-ju replied:
Updated feedback component to use govuk-frontend classes for layout in 6ab1174

andysellick commented on app/assets/stylesheets/govuk_publishing_components/components/_button.scss:
```diff
@@ -110,7 +110,7 @@ $gem-quiet-button-hover-colour: darken($gem-quiet-button-colour, 5%);
 // Begin adjustments for font baseline offset
 // These should be removed when the font is updated with the correct baseline
-// For the 1px addition please see https://github.com/alphagov/govuk-frontend/pull/365#discussion_r154349428
+// For the 1px addition please see https://github.com/alphagov/govuk/pull/365#discussion_r154349428
```
Not sure this one should have been changed from govuk-frontend to govuk, possibly a victim of find and replace?

alex-ju replied:
The initial replace was 'erroneous', but the adjustments (and the comments) were removed in a following commit.

andysellick commented on app/assets/stylesheets/govuk_publishing_components/components/_checkboxes.scss:
```diff
@@ -27,7 +27,7 @@
 .gem-c-checkboxes {
   // this is complex but needed to override the govuk-frontend styles in
-  // https://github.com/alphagov/govuk-frontend/blob/b8058640b9602ecb6e1f66f887553190cbae7b46/src/components/hint/_hint.scss#L16
+  // https://github.com/alphagov/govuk/blob/b8058640b9602ecb6e1f66f887553190cbae7b46/src/components/hint/_hint.scss#L16
```
Also not sure this one should have been changed from govuk-frontend to govuk.

alex-ju replied:
Fixed, missed when replacing. Good spot.

andysellick commented on config/initializers/assets.rb:
```diff
@@ -34,7 +34,8 @@
-  #{__dir__}/../../node_modules/govuk-frontend/assets/fonts
+  #{__dir__}/../../node_modules/govuk-frontend/govuk/assets/images
+  #{__dir__}/../../node_modules/govuk-frontend/govuk/assets/fonts
+  #{__dir__}/../../node_modules/govuk-frontend/
```
Question: if the whole of the `node_modules/govuk-frontend` dir is a path, why do we need the images and fonts subdirs explicitly specified?

alex-ju replied:
I suspect this is because they're referenced without the full path (e.g. `my-font.otf` instead of `/govuk/assets/fonts/my-font.otf`)

andysellick replied:
I've tested this and those extra paths are definitely needed. Weird. Oh well.

andysellick commented on app/assets/stylesheets/govuk_publishing_components/components/_feedback.scss:
```diff
@@ -33,10 +33,12 @@
 
 .gem-c-feedback__column-two-thirds {
-  @include govuk-grid-column(two-thirds, $class: false)
+  @include govuk-grid-column(two-thirds)
```
Nitpick: no semicolon at the end of this line (although presumably it passed the linter?).

alex-ju replied:
This code was removed in a following commit.

andysellick commented on app/assets/stylesheets/govuk_publishing_components/components/_feedback.scss:
```diff
@@ -33,10 +33,12 @@
 
-@include govuk-grid-row($class: "gem-c-feedback__grid-row");
+.gem-c-feedback__grid-row {
+  @extend %govuk-grid-row;
```
V3 changelog says `You can no longer use the govuk-grid-row mixin` - that's not this, is it? I haven't checked the review app (it's not working atm).

alex-ju replied:
Yes, it's that one. This was removed in a following commit.

andysellick commented on app/assets/stylesheets/govuk_publishing_components/components/_modal-dialogue.scss:
```diff
@@ -136,7 +138,10 @@ $govuk-modal-wide-breakpoint: $govuk-page-width + $govuk-modal-margin * 2 + $gov
+    @include govuk-focused-text;
+
     outline: none;
+    box-shadow: none;
```
Probably outside of scope, but I think the close button looks pretty good with the box shadow. Certainly helps with the contrast between its focussed state and the white space below.

With:

<img width="741" alt="Screen Shot 2019-08-05 at 17 23 17" src="https://user-images.githubusercontent.com/861310/62479617-c9860080-b7a5-11e9-85f6-1513fe4f8356.png">

Without:

<img width="743" alt="Screen Shot 2019-08-05 at 17 23 09" src="https://user-images.githubusercontent.com/861310/62479603-becb6b80-b7a5-11e9-8796-f46500551906.png">

alex-ju replied:
Good call for potential improvement

alex-ju replied:
Added a commit to improve the focus on the modal's close button
<img width="1039" alt="Screen Shot 2019-08-07 at 11 11 17" src="https://user-images.githubusercontent.com/788096/62614871-25ff3200-b904-11e9-8258-7d0b4f604d55.png">

andysellick commented on app/assets/stylesheets/govuk_publishing_components/components/_search.scss:
```diff
@@ -30,14 +29,22 @@ $large-input-size: 50px;
+
+  &:focus {
+    outline: $govuk-focus-width solid $govuk-focus-colour;
+    outline-offset: inherit;
```
Probably out of scope, but the search box is long overdue an overhaul to make it look like GOVUK Frontend. We should talk about that.

alex-ju replied:
I agree about this needs an overhaul. Again, good call for an improvement.

andysellick replied:
I think we should address this after V3 has been merged. Issue logged here: https://github.com/alphagov/govuk_publishing_components/issues/1027

andysellick commented on app/assets/stylesheets/govuk_publishing_components/components/_step-by-step-nav.scss:
```diff
@@ -237,7 +237,9 @@ $top-border: solid 2px $grey-3;
-    @include govuk-focusable;
+    &:focus {
+      outline: $govuk-focus-width solid $govuk-focus-colour;
+    }
```
Not sure what's happened here but the show/hide all button on step navs (which has this class) seems to have lost its focus state.

<img width="802" alt="Screen Shot 2019-08-05 at 17 29 14" src="https://user-images.githubusercontent.com/861310/62480028-af98ed80-b7a6-11e9-9888-bc00b03861f7.png">

nickcolley replied:
Perhaps one for a follow up pull request but that style will not pass non-text guidelines.

https://design-system.service.gov.uk/get-started/focus-states/

alex-ju replied:
I did an update on the step-by-step-nav to improve focus states.

<img width="960" alt="Screen Shot 2019-08-07 at 11 08 42" src="https://user-images.githubusercontent.com/788096/62614689-cef95d00-b903-11e9-9caa-002b2fc45536.png">

This component is also a good candidate for an overhaul looking at the use of `!important`

nickcolley replied:
Looking good, however there looks like some issues around hover + focus:

<img width="166" alt="Screen Shot 2019-08-07 at 13 46 59" src="https://user-images.githubusercontent.com/2445413/62623959-f1967080-b919-11e9-8613-cc5f083dff8f.png">

alex-ju replied:
Yep, I'm thinking how to tackle that by untangling styles rather than chaining selectors.

alex-ju commented on app/views/govuk_publishing_components/components/feedback/_survey_signup_form.html.erb:
```diff
@@ -12,7 +12,7 @@
     role="button">Close</a>
 
-  <div class="gem-c-feedback__grid-row gem-c-feedback__survey-container">
+  <div class="govuk-grid-row">
```
I think this was re-introduced at rebasing. The padding is unnecessary from my point of view too, except for the home page where the layout should provide this padding on mobile (this was the reason for introducing it recently if I recall well).

alex-ju replied:
Removed the unused class.

36degrees commented on app/views/govuk_publishing_components/components/_layout_header.html.erb:
```diff
@@ -36,7 +36,7 @@
 
     <% if navigation_items.any? %>
-    <button role="button" class="govuk-header__menu-button js-header-toggle" aria-controls="navigation" aria-label="Show or hide Top Level Navigation">Menu</button>
+    <button role="button" class="govuk-header__menu-button govuk-js-header-toggle" aria-controls="navigation" aria-label="Show or hide Top Level Navigation">Menu</button>
```
This is a change from GOV.UK Frontend – we [introduced namespacing of JS classes](https://github.com/alphagov/govuk-frontend/pull/1444) to fix an [issue using the header component with GOV.UK Template](https://github.com/alphagov/govuk-frontend/issues/1270).

36degrees commented on app/assets/stylesheets/govuk_publishing_components/components/helpers/_variables.scss:
```diff
@@ -11,7 +11,7 @@ $gem-spacing-scale-7: 50px;
 
 $gem-text-colour: $text-colour;
-$gem-secondary-text-colour: $grey-1;
+$gem-secondary-text-colour: govuk-colours("dark-grey");
```
[Aside: This'll compile just fine because Sass will assume that `govuk-colours` is a native CSS construct (like e.g. `background-image`) – this sort of thing can easily slip in, so we've written [a test to try and catch it](https://github.com/alphagov/govuk-frontend/blob/08b408840deabc40cb9d2438b0d19862cda41f05/src/govuk/all.test.js#L116-L155)]

alex-ju replied:
Ha, good found, and even better test, but it's a bit hard to replicate it in the world of Jasmine. The mixin failed silently and it had no visual effects since this variable is no longer used (was replaced by `$govuk-secondary-text-colour` all over the gem. This should be removed now.

andysellick commented:
Some more thoughts...

- the accessible autocomplete doesn't have the new focus styles (because it's mostly pulled in from another repo) but we're going to address that separately
- links that wrap have a potential overlap issue with the line below slightly cropping the line above, as seen in the attachment component (note how the bottom of the 'P' in 'department' is missing). Might be something to talk to the Design System team about (also we might need to adjust the line height in this component)

<img width="583" alt="Screen Shot 2019-08-06 at 13 21 54" src="https://user-images.githubusercontent.com/861310/62539524-46b88080-b84d-11e9-9417-b8038818b651.png">

Here's an only semi-related but interesting example from the contents list component, where the second line is indented, leaving a trailing bit of the first line's underline visible (I thought it was a quote mark when I first spotted it)

<img width="749" alt="Screen Shot 2019-08-06 at 13 48 53" src="https://user-images.githubusercontent.com/861310/62541103-fe9b5d00-b850-11e9-8f97-4390eaf9844e.png">

andysellick commented:
Also fairly obvious but worth pointing out that apps will break with this update - I just tried it in finder-frontend and it failed because apparently we're using the `govuk-focusable` mixin in that app.

alex-ju commented on app/views/govuk_publishing_components/components/_textarea.html.erb:
```diff
@@ -20,7 +20,7 @@
 
   css_classes = %w(govuk-textarea)
-  css_classes << "js-character-count" if character_count
+  css_classes << "govuk-js-character-count" if character_count
```
It's not unnecessary. We used to have a clash on `js-header-toggle` being picked up by 2 scripts. I needed to find a workaround for that in the past, using a more explicit selector. Now these can co-exist until we move to the new header component.

andysellick commented:
Tables look good with the new, slightly lighter grey...

![Screen Shot 2019-08-07 at 10 21 48](https://user-images.githubusercontent.com/861310/62611318-3fe94680-b8fd-11e9-8afd-419536d630ad.png)

Old:

![Screen Shot 2019-08-07 at 10 21 42](https://user-images.githubusercontent.com/861310/62611331-45469100-b8fd-11e9-8163-1672a3ac31e2.png)

andysellick commented:
The new focus style takes up a bit more vertical space than it used to. The worst example of this so far that I've found is the image card (also noticeable with the textarea, with text beneath it).

<img width="796" alt="Screen Shot 2019-08-07 at 11 06 00" src="https://user-images.githubusercontent.com/861310/62614511-7aee7880-b903-11e9-8c3c-1f7be8d6a0d5.png">

I don't know if this is actually a problem though. You're most likely to notice this if you're using a keyboard, in which case moving the focus won't be a problem. Also the one you're currently focussed on is fine.

Might be worth looking at some of these issues after this PR is merged, just noting it here for reference.

DilwoarH commented on app/assets/stylesheets/component_guide/application.scss:
```diff
@@ -2,12 +2,10 @@
-@import "govuk-frontend/helpers/all";
-@import "govuk-frontend/core/all";
-
-@include _govuk-font-face-nta;
```
does this no longer need importing?

alex-ju replied:
Nope, that was a hack to use the new font in the component guide app as it wasn't loaded because of the compatibility flags. Now that we have them turned off the font comes by default and this is no longer needed.

andysellick commented:
This is unrelated but worth adding to the 'to check later' list... the layout header with links doesn't seem to collapse early enough, plus when the links do collapse the dropdown menu doesn't seem to work?

<img width="784" alt="Screen Shot 2019-08-07 at 13 14 28" src="https://user-images.githubusercontent.com/861310/62622091-6fa44880-b915-11e9-9e00-24a055a4eb30.png">

andysellick commented:
Page title (with context link) lacks a focus state for the link. This was a problem prior to this PR, so arguably out of scope, but we should fix it.

<img width="779" alt="Screen Shot 2019-08-07 at 13 27 10" src="https://user-images.githubusercontent.com/861310/62622818-38369b80-b917-11e9-890b-b3fc332cb564.png">

vanitabarrett commented:
Sorry if I missed this in our review meeting of this earlier - it looks like some components have been updated to use the new focus style, but some haven't. What is the reason behind that? 

E.g:

<img width="988" alt="Screen Shot 2019-08-07 at 14 22 17" src="https://user-images.githubusercontent.com/29889908/62626214-c6fae680-b91e-11e9-9771-f890583fcb17.png">

alex-ju commented on app/assets/stylesheets/govuk_publishing_components/components/govspeak/_charts.scss:
```diff
@@ -26,7 +26,7 @@
   $chart-border: govuk-colour("white"); // Chart border colour
   $key-border: govuk-colour("white"); // Key border colour
-  $bar-colours: govuk-colour("blue"), govuk-colour("turquoise"), govuk-colour("green"), govuk-colour("light-green"), govuk-colour("yellow"), govuk-colour("orange"), govuk-colour("red"), govuk-colour("bright-red");
+  $bar-colours: govuk-colour("blue"), govuk-colour("turquoise"), govuk-colour("green"), govuk-colour("light-green"), govuk-colour("yellow"), govuk-colour("orange"), govuk-colour("red"), govuk-colour("red");
```
Yes, that's right. We could use purple, or any other [colour from the palette](https://design-system.service.gov.uk/styles/colour/#colour-palette). Open to suggestions 🎨

alex-ju replied:
Updated to use `bright-purple` for now, the closest in shade with the previous one.

vanitabarrett commented:
Do we need to update the govspeak 'highlight answer' to use the green instead of the turquoise?
https://govuk-publishing-compo-pr-1010.herokuapp.com/component-guide/govspeak/highlight_answer/preview

<img width="946" alt="Screen Shot 2019-08-07 at 14 50 02" src="https://user-images.githubusercontent.com/29889908/62628158-af256180-b922-11e9-81e9-616775f09cb0.png">

andysellick commented:
The hover + focus state for next/previous navigation could do with a bit of attention...

Focus:

![Screen Shot 2019-08-07 at 14 53 23](https://user-images.githubusercontent.com/861310/62628421-34a91180-b923-11e9-8b42-fe481e67719c.png)

Focus + hover:

![Screen Shot 2019-08-07 at 14 53 29](https://user-images.githubusercontent.com/861310/62628438-3e327980-b923-11e9-9c84-6cc27954b002.png)

andysellick commented:
Not sure we can do much about this one and it predates this PR, but focussed share links with branding applied might not always pass contrast requirements. At some point we should check.

![Screen Shot 2019-08-07 at 16 05 28](https://user-images.githubusercontent.com/861310/62634206-61fabd00-b92d-11e9-80b5-e56259756cc9.png)

vanitabarrett commented:
@andysellick I did look through those as I was concerned about focus states, but I think they all pass colour contrast (AA)

andysellick commented:
Skip link hasn't picked up the new focus style, might be defining it itself?

![Screen Shot 2019-08-07 at 16 08 45](https://user-images.githubusercontent.com/861310/62634384-b56d0b00-b92d-11e9-88fb-c075360894c6.png)

andysellick commented:
Step by step navigation context text fails contrast requirements when focussed (probably did before this PR).

![Screen Shot 2019-08-07 at 16 11 55](https://user-images.githubusercontent.com/861310/62634639-257b9100-b92e-11e9-87a1-1caec66ff164.png)

andysellick commented:
We might want to tweak the image card component a little... if you focus and hover some of the links, you get a double underline.

![Screen Shot 2019-08-07 at 16 18 11](https://user-images.githubusercontent.com/861310/62635141-0df0d800-b92f-11e9-8d77-9ebd93a49e1e.png)

The component overrides the default `text-decoration: none` on hover + focus from govuk-frontend to add underline. We should remove it in this specific case i.e. leave underline on when hovering.

andysellick commented:
Again seems to be unrelated but the sort option for the table component seems to be non-functional - clicking it goes back to the guide homepage (pointing at /).

<img width="801" alt="Screen Shot 2019-08-07 at 16 52 41" src="https://user-images.githubusercontent.com/861310/62637678-e6e8d500-b933-11e9-8688-8c71b457e782.png">

@alex-ju once I've finished going through the components I'll create issues for anything I've raised here that shouldn't be fixed as part of this PR.

andysellick commented on app/assets/stylesheets/govuk_publishing_components/components/_modal-dialogue.scss:
```diff
@@ -97,7 +97,7 @@ $govuk-modal-wide-breakpoint: $govuk-page-width + $govuk-modal-margin * 2 + $gov
 
 .gem-c-modal-dialogue__header {
-  padding: 8px govuk-spacing(3) 10px;
+  padding: 9px govuk-spacing(3) 10px;
```
I don't really mind, but why have an explicit `px` value alongside a `govuk-spacing` value?

alex-ju replied:
Yeah, it's an odd mix isn't it. I can use pixels for all values for consistency. Tried to make sure the vertical rhythm is preserved, that's why I always use `govuk-spacing` for top and bottom.

andysellick commented:
@alex-ju I've logged issues for anything I've found that's unrelated to this PR and marked those comments accordingly. Some of them should be addressed after this PR is merged and some can be worked on right now - I'll take a look at some of those when I have a chance.

andysellick commented on app/assets/stylesheets/govuk_publishing_components/components/_table.scss:
```diff
@@ -1,15 +1,15 @@
 $sort-link-arrow-spacing: $sort-link-arrow-size / 2;
 $table-row-hover-background-colour: rgba(43, 140, 196, .2);
-$table-row-even-background-colour: govuk-colour("grey-4");
+$table-row-even-background-colour: govuk-colour("light-grey");
```
I've gone through all the components and haven't spotted any problems relating to this.

alex-ju commented:
@vanitabarrett Thanks for the review. Tried to address the comments. Do let me know if the review is still in progress.
> Do we need to update the govspeak 'highlight answer' to use the green instead of the turquoise?
> https://govuk-publishing-compo-pr-1010.herokuapp.com/component-guide/govspeak/highlight_answer/preview

Updated to use the new colour

@andysellick Thanks for the review. Tried to address all the comments (aside from the ones with an associated issue - 🙌 for that). For those comments without a thread I tried to reply below.

> The hover + focus state for next/previous navigation could do with a bit of attention...

Updated to preserve the focus state when hovering

> Skip link hasn't picked up the new focus style, might be defining it itself?

This is how the component is coming from `govuk-frontend`. We can propose enhancements upstream or do overwrites if we have good reasons for that.

> We might want to tweak the image card component a little... if you focus and hover some of the links, you get a double underline.

Fixed the focus/hover states on image card component

alex-ju commented:
Rebased again before merging. Thanks everyone who commented on this, that helped a lot!
